daemon/server/listen.go

```diff
@@ -0,0 +1,87 @@
+package server
+
+import (
+	"net"
+	"os"
+	"path"
+	"syscall"
+)
+
+func listen(socketPath string) (net.Listener, error) {
+	err := os.MkdirAll(path.Dir(socketPath), 0700)
+	if err != nil {
+		return nil, err
+	}
+	l, err := net.Listen("unix", socketPath)
+	if err == nil {
+		return l, nil
+	}
+
+	l = replaceDeadServer(socketPath, err)
+	if l != nil {
+		return l, nil
+	}
+	return nil, err
+}
+
+// replaceDeadServer handles the common case where a daemon has died but the socket file still exists.
+// If the address is already in use, we try connecting to it.
+// If we get connection refused, we infer the server is dead and
+// remove the socket file, and then try serving.
+// Returns a valid listener or nil.
+func replaceDeadServer(socketPath string, err error) net.Listener {
+	if !isAddrAlreadyInUse(err) {
```

Comment:
I'd clarify why InUse is the only valid case for proceeding with replacement (because we consider every other error to be unrecoverable and to require manual intervention).